parcels/interaction/hash_flat.py

```diff
@@ -0,0 +1,145 @@
+import numpy as np
+from parcels.interaction.base_neighbor import BaseFlatNeighborSearch
+# from numba import njit
+# import numba as nb
+from parcels.interaction.base_hash import BaseHashNeighborSearch, hash_split
+
+
+class HashFlatNeighborSearch(BaseHashNeighborSearch, BaseFlatNeighborSearch):
+    '''Neighbor search using a hashtable (similar to octtrees).'''
+    name = "hash"
+    _box = None
+
+    def _find_neighbors(self, hash_id, coor):
+        neighbor_blocks = hash_to_neighbors(hash_id, self._bits)
+        all_neighbor_points = []
+        for block in neighbor_blocks:
+            try:
+                all_neighbor_points.extend(self._hashtable[block])
+            except KeyError:
+                pass
+
+        pot_neighbors = np.array(all_neighbor_points)
+        return self._get_close_neighbor_dist(coor, pot_neighbors)
+
+    def update_values(self, new_values, new_active_mask=None):
+        if not self._check_box(new_values, new_active_mask):
+            self.rebuild(new_values, new_active_mask)
+        else:
+            super().update_values(new_values, new_active_mask=new_active_mask)
+
+    def _check_box(self, new_values, new_active_mask):
+        """Check whether particles have moved out of the overall box.
+
+        :param new_values: New particle coordinates to be checked.
+        :param new_active_mask: New active mask for the particles.
+        :returns True if box is still big enough, False if not.
+        """
+        if self._box is None:
+            return False
+        if new_active_mask is None:
+            active_values = new_values
+        else:
+            active_values = new_values[:, new_active_mask]
+        for i_dim in range(3):
+            if np.any(active_values[i_dim, :] - self._box[i_dim][0] < 0):
+                return False
+            if np.any(active_values[i_dim, :] - self._box[i_dim][1] > 0):
+                return False
+        return True
+
+    def rebuild(self, values, active_mask=-1):
+        super().rebuild(values, active_mask)
+        active_values = self._values[:, self._active_mask]
+
+        # Compute the dimensions of the box with a margin.
+        self._box = []
+        for i_dim in range(3):
+            val_min = active_values[i_dim, :].min()
+            val_max = active_values[i_dim, :].max()
+            margin = (val_max-val_min)*0.3
+            self._box.append([val_min-margin, val_max+margin])
+
+        self._box = np.array(self._box)
+
+        epsilon = 1e-8
+
+        # Compute hash tables and the bits to compute it.
+        n_bits = ((self._box[:, 1] - self._box[:, 0]
+                   )/self.inter_dist.reshape(-1) + epsilon)/np.log(2)
+        self._bits = np.ceil(n_bits).astype(int)
```

Comment:
It should not allocate the same bit-budget for each dimension actually: self._bits is an array with 3 values (one for each dimension. Perhaps needs more documentation/comments? (Since these values get hashed, it's not actually that bad to be wasteful.)

Reply:
well, if you can perhaps just make a short comment about it that would suffice - at the location of the code. Thanks also for highlighting & explaining the `self._bits` scheme.

Reply:
I have added some comments.